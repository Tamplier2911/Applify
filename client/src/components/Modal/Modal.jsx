// import "./Modal.scss";
import React from "react";
import ReactDOM from "react-dom";

// redux
import { connect } from "react-redux";
import { createStructuredSelector } from "reselect";
import { closeModal } from "../../redux/modal/modal.actions";
import {
  selectModalHidden,
  selectModalContent,
} from "../../redux/modal/modal.selectors";

// JS Rendering CSS
import {
  ModalContainer,
  ModalContentBox,
  ModalHeader,
  ModalContent,
  ModalClose,
} from "./ModalStyles";

/**
 * Modal - default modal takes modal condition and modal content from state.
 * @param {boolean} modalCondition - condition to show modal.
 * @param {object} modalContent - contains header and content strings.
 * @param {boolean} test - since it is quite complicated perform testing on portals,
 * if test property is passed to component -> markdown will be returned, else portal.
 */

const Modal = ({ modalCondition, modalContent, closeModal, test }) => {
  const { header, content } = modalContent;

  const markdown = (
    <ModalContainer
      data-test="modal"
      onClick={closeModal}
      hidden={modalCondition}
    >
      <ModalContentBox onClick={(e) => e.stopPropagation()}>
        <ModalHeader data-test="modal-header">{header}</ModalHeader>
        <ModalClose data-test="modal-close" onClick={closeModal}>
          &#10006;
        </ModalClose>
        <ModalContent data-test="modal-content">{content}</ModalContent>
      </ModalContentBox>
    </ModalContainer>
  );

  return test
    ? markdown
    : ReactDOM.createPortal(markdown, document.querySelector("#modal"));
};

const mapStateToProps = createStructuredSelector({
  modalCondition: selectModalHidden,
  modalContent: selectModalContent,
});

export default connect(mapStateToProps, { closeModal })(Modal);
