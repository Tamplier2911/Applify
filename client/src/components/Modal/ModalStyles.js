import styled, { css } from "styled-components";

const showModalAnimation = css`
  @keyframes showModal {
    0% {
      opacity: 0;
    }
    100% {
      opacity: 1;
    }
  }
`;

const dysplayModal = props => {
  const { hidden } = props;
  if (hidden) {
    return `
        display: none
    `;
  } else {
    return `
        display: grid;
        font-family: var(--font-title);
        position: fixed;
        width: 100vw;
        height: 100vh;
        background-color: var(--cl-sidenav);
        z-index: 150;
        
        align-items: center;
        justify-items: center;
        
        @media only screen and (max-width: 310px) {
            width: 300vw;
            height: 300vh;
            display: inline-block;
        }
    `;
  }
};

export const ModalContainer = styled.div`
  ${dysplayModal}
`;

export const ModalContentBox = styled.div`
  ${showModalAnimation}

  display: grid;
  grid-template-rows: 2fr 0.5fr 7fr;

  position: relative;
  grid-auto-rows: min-content;
  width: 40rem;
  height: 30rem;

  padding: 2rem;
  overflow-x: hidden;
  overflow-y: scroll;

  background-color: var(--cl-white);
  box-shadow: 0 0.2rem 0.4rem var(--cl-sidenav-shadow);

  animation: showModal 1s;

  @media only screen and (max-width: 360px) {
    width: 30rem;
    height: 40rem;
  }

  @media only screen and (max-width: 310px) {
    margin: 10rem auto auto 4rem;
  }

  @media only screen and (max-width: 300px) {
    margin: 0;
    width: 100vw;
    height: 100vh;
  }

  @media only screen and (max-width: 200px) {
    width: 100vw;
    height: 130vh;
  }

  &::-webkit-scrollbar {
    width: 0.4rem;
    background-color: var(--cl-scroll-bg);
  }

  &::-webkit-scrollbar-thumb {
    background-image: linear-gradient(
      to bottom,
      var(--cl-scrollThumb-bg1),
      var(--cl-scrollThumb-bg2)
    );
    border-radius: 5rem;
  }

  &::-webkit-scrollbar-track {
    -webkit-box-shadow: inset 0 0 6px var(--cl-scroll-shadow);
    box-shadow: inset 0 0 6px var(--cl-scroll-shadow);
  }

  &::after {
    content: "";
    margin: 0.4rem 0 0.4rem 0;
    grid-row: 2 / 3;
    display: inline-block;
    width: 100%;
    height: 0.2rem;
    background-color: var(--cl-dark-grey);
  }
`;

export const ModalHeader = styled.h3`
  font-size: 2.4rem;
  font-weight: 400;
`;

export const ModalClose = styled.div`
  position: absolute;
  cursor: pointer;
  top: 2rem;
  right: 2rem;
  font-size: 2rem;
  transition: color 0.3s;

  &:hover,
  &:active {
    color: var(--cl-primary);
  }
`;

export const ModalContent = styled.div`
  font-size: 1.7rem;
`;
