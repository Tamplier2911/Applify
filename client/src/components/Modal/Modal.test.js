import React from "react";
import toJson from "enzyme-to-json";

// component
import Modal from "./Modal";

// utils
import {
  setShallowC,
  findByTestAttr,
  storeFactory,
} from "../../tests/testUtils";

describe("<Modal />", () => {
  const initialStore = {
    modal: {
      modalHidden: false,
      modalContent: {
        header: "Modal",
        content: "Testing modal!",
      },
    },
  };
  const store = storeFactory(initialStore);

  let wrapper;
  beforeEach(() => {
    wrapper = setShallowC(<Modal store={store} test={true} />);
  });

  it("renders without an error", () => {
    const modal = findByTestAttr(wrapper, "modal");
    expect(modal.length).toBe(1);
  });

  it("renders expected content from state", () => {
    const modalHeader = findByTestAttr(wrapper, "modal-header");
    expect(modalHeader.length).toBe(1);
    expect(modalHeader.text()).toBe("Modal");
    const modalContent = findByTestAttr(wrapper, "modal-content");
    expect(modalContent.length).toBe(1);
    expect(modalContent.text()).toBe("Testing modal!");
  });

  it("closes when closeModal action button is clicked", () => {
    const closeModal = findByTestAttr(wrapper, "modal-close");
    expect(closeModal.length).toBe(1);
    closeModal.simulate("click");
    const {
      modal: { modalHidden },
    } = store.getState();
    expect(modalHidden).toBeTruthy();
  });

  it("matches snapshot", () => {
    const modal = findByTestAttr(wrapper, "modal");
    expect(toJson(modal)).toMatchSnapshot();
  });
});
