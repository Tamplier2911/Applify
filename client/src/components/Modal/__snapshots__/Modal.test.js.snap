// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Modal /> matches snapshot 1`] = `
<styled.div
  data-test="modal"
  hidden={true}
  onClick={[Function]}
>
  <styled.div
    onClick={[Function]}
  >
    <styled.h3
      data-test="modal-header"
    />
    <styled.div
      data-test="modal-close"
      onClick={[Function]}
    >
      âœ–
    </styled.div>
    <styled.div
      data-test="modal-content"
    />
  </styled.div>
</styled.div>
`;
